File: 2021/d1/main.pl
#!/usr/bin/perl

use strict;
use warnings;

my @input = <>;

sub part1 {
	my $last = 999999;
	my $incs = 0;
	for(@input){
		chomp(my $n = $_);
		if($n > $last){
			$incs++;
		}
		$last = $n;
	}
	print "Part 1: $incs\n";
}

sub part2 {
	my $last = 99999999;
	my $incs = 0;

	for(my $i = 0; $i < @input; $i++){
		if($i+2 >= @input){
			last;
		}
		my @window = ($input[$i], $input[$i+1], $input[$i+2]);
		my $winsum = $window[0]+$window[1]+$window[2];

		if($winsum > $last){
			$incs++;
		}
		$last = $winsum;
	}
	print "Part 2: $incs\n";
}

part1;
part2;

File: 2021/d1/src/main.rs
fn main() -> Result<(), Box<dyn std::error::Error>> {
    let input = std::fs::read_to_string("input.txt")?
        .split('\n')
        .filter(|l| !l.is_empty())
        .map(str::parse)
        .collect::<Result<Vec<u32>, _>>()?;

    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));

    Ok(())
}

fn part1(input: &Vec<u32>) -> u32 {
    input
        .iter()
        .copied()
        .fold((0, None), |(incs, last), depth| match last {
            None => (0, Some(depth)),
            Some(last) => {
                if depth > last {
                    (incs + 1, Some(depth))
                } else {
                    (incs, Some(depth))
                }
            }
        })
        .0
}

fn part2(input: &Vec<u32>) -> u32 {
    input
        .windows(3)
        .fold((0, None), |(incs, last): (u32, Option<u32>), depths| {
            let depth = depths.iter().sum();

            match last {
                None => (0, Some(depth)),
                Some(last) => {
                    if depth > last {
                        (incs + 1, Some(depth))
                    } else {
                        (incs, Some(depth))
                    }
                }
            }
        })
        .0
}
